jetstream/core/orchestrator.py

```diff
@@ -568,8 +569,9 @@ def _detokenize_thread(self, idx: int):
     my_slots = self._generate_slots[idx]
 
     metadata = my_generate_engine.get_tokenizer()
-    vocab = token_utils.load_vocab(metadata.path, metadata.extra_ids)
-
+    tokenizer = my_generate_engine.build_tokenizer(metadata)
+    if not tokenizer:
```

Comment:
If we let build_tokenizer handle build tokenizer, lest's give engine full responsibility of build_tokenizer.  Engine implementation call token_utils.SentencePieceTokenizer if needed. The error are expected if tokenizer is still None after build_tokenizer. 

Same as line 400 to 402

Reply:
I agree with your point but changing the build_tokenizer default implementation to return token_utils.SentencePieceTokenizer will cause a circular import issue
engine_api -> token_utils -> engine_api. For now, I think it would be better to keep it this way and then I can update the MaxText engine to return a SentencePiece Tokenizer.

Reply:
Updated it remove the circular import by using Any typing annotation.

Reply:
Great! Thanks for the change and fixing the circular import issues!